Kcena 6주차 과제1 [레스토랑 상세 페이지 구현하기] 제출

Kwakcena opened the pull request:
## 6주차 과제1 [레스토랑 상세 페이지 구현하기] 제출합니다. 

### 과제를 통해 알아야 할 것.
- `useEffect` 의 동작이 어떻게 되는가?
- react 컴포넌트는 어떤 순서로 렌더링 되는가?
- `base path`는 무엇이고, 왜 index.html에서 `/main.js' 를 써야 잘 동작하는가?
- `api` 테스트와 `비동기 action` 테스트는 어떻게 작성된 것인가 ?

Kwakcena committed:
HomePage 테스트

HomePage 컴포넌트를 렌더링 하면 타이틀과 클릭할 수 있는 링크들이
나타난다.

Kwakcena committed:
HomePage 테스트 구현

Kwakcena committed:
Refactor: App에서 HomePage 사용

Kwakcena committed:
RestaurantsPage로 분리하기

레스토랑의 지역, 분류 선택을 담당하는 페이지를 컴포넌트로 따로 분리
했다.

Kwakcena committed:
App에서 RestaurantsPage 사용

Kwakcena committed:
react-router-dom 설치

Kwakcena committed:
About 페이지 접속 테스트

App에서 About을 클릭 시 /about 페이지에 접속하는지 테스트.

Kwakcena committed:
Router 사용하기 위한 BrowserRouter 사용

Kwakcena committed:
테스트 코드 수정

Kwakcena committed:
About 페이지 테스트 통과

Kwakcena committed:
AboutPage 테스트

Kwakcena committed:
AboutPage 구현

Kwakcena committed:
Refactor: AboutPage 컴포넌트 사용

Kwakcena committed:
App에서 render 할 때 메인 페이지 보이는 테스트

처음 App에 접속하면 메인 페이지가 보인다.

Kwakcena committed:
메인 페이지 접속 테스트 구현

Kwakcena committed:
Refactor: 중복 제거, fixtures 사용

Kwakcena committed:
레스토랑 목록 페이지 테스트

Kwakcena committed:
테스트 코드 수정

fixtures 데이터의 지역, 카테고리, 레스토랑 목록을 그대로 사용하도록 수정

Kwakcena committed:
Restaurants 접속 구현

Kwakcena committed:
webpack config update

웹팩 설정에서 잘못된 페이지 접속 시 리다이렉션을 설정한다.

Kwakcena committed:
HomePage에서 Link를 사용

Link를 사용하기 위해서 MemoryRouter로 묶어준다.

Kwakcena committed:
Link 추가 구현

Kwakcena committed:
NotFoundPage 테스트

Kwakcena committed:
NotFoundPage 구현

Kwakcena committed:
NotFound 페이지 App test에 추가

Kwakcena committed:
NotFound 페이지 App에 구현

Kwakcena committed:
헤더 클릭 시 메인 페이지 테스트

헤더를 클릭하면 메인페이지로 돌아가는 테스트

Kwakcena committed:
헤더 클릭시 메인 페이지 구현

Kwakcena committed:
fixtures restaurants 정보 수정

Kwakcena committed:
Restaurant 테스트

하나의 레스토랑에 대한 컴포넌트 테스트

Kwakcena committed:
Restaurant 구현

Kwakcena committed:
Refactor: Restaurant 컴포넌트 사용

Kwakcena committed:
레스토랑 fixture 추가

Kwakcena committed:
Refactor: fixtures 사용

Kwakcena committed:
Rename Restaurant

Kwakcena committed:
Link 사용을 위해 MemoryRouter로 묶어줌

Kwakcena committed:
Link 사용

Kwakcena committed:
RestaurantContainer 삭제

Kwakcena committed:
Refactor: 보기좋게 수정

Kwakcena committed:
접속 url 수정

Kwakcena committed:
레스토랑 하나의 정보를 보여주는 페이지

레스토랑 하나의 정보를 보여주는 페이지 컴포넌트

Kwakcena committed:
test 코드 수정

Kwakcena committed:
RestaurantPage 구현

Kwakcena committed:
RestaurantPage 적용

Kwakcena committed:
잘못된 코드 수정

Kwakcena committed:
RestaurantContainer 테스트

선택한 레스토랑 정보를 불러오는 컨테이너 컴포넌트 테스트

Kwakcena committed:
RestaurantContainer 하드코딩으로 통과

Kwakcena committed:
Refactor: RestaurantContainer 사용

Kwakcena committed:
setRestaurat reducer test

레스토랑 선택시 reducer 작동 테스트

Kwakcena committed:
setRestaurant 구현

hannut91 commented:
잘 진행하고 계신 것 같아요 ㅎㅎ 계속 진행해주세요 😀

Kwakcena committed:
setRestaurantId 테스트

../restaurants/1 과 같은 url에서 id에 해당하는 1을 match로 뽑아낼 수
있으므로 restaurantPage 컴포넌트가 실행되면 dispatch로 id를 전달해서
setRestaurantId로 id를 저장할 계획.

저장한 뒤 loadRestaurant 액션을 통해 getState로 id를 읽어올 수 있으므로
이를 fetch 에 전달하여 해당 레스토랑 정보를 읽을 수 있다.

Kwakcena committed:
setRestaurantId 구현

Kwakcena committed:
RestaurantContainer 테스트

레스토랑 컨테이너 컴포넌트는 선택한 레스토랑 정보를 useSelector를 이용해
읽어온다.

Kwakcena committed:
RestaurantContainer 구현

useSelector를 적용

Kwakcena committed:
Refactor: RestaurantContainer 리팩토링

utils의 get 사용
레스토랑 정보 출력

Kwakcena committed:
RestaurantPage dispatch 호출 테스트

Kwakcena committed:
RestaurantPage dispatch 호출 테스트 구현

Kwakcena committed:
loadRestaurant action 테스트

Kwakcena committed:
코드 수정

Kwakcena committed:
loadRestaurant, fetchRestaurant 구현

Kwakcena committed:
loadRestaurant action 테스트 추가

restaurantId가 없을 경우 테스트 추가

Kwakcena committed:
api test 추가

Kwakcena committed:
reducer test refactor

Kwakcena committed:
reducer 초기값 수정

Kwakcena committed:
RestaurantPage에 match props 전달해주기

Kwakcena committed:
match 추가 테스트 구현

Kwakcena committed:
RestaurantPage test 리팩토링

Kwakcena committed:
RestaurantContainer 리팩토링

li에 key 할당

Kwakcena committed:
useEffect 빈 배열  추가

Kwakcena commented:
지역과 분류를 선택하여 나온 레스토랑 이름 목록에 `Link` 태그를 걸었습니다.

이를 클릭하면 해당 레스토랑의 id가 url에 포함된 주소로 이동합니다. 예를 들어 다음과 같습니다. `restaurants/1` 

이에 대한 `Route` 는 App 컴포넌트에서 `<Route path="/restaurants/:id" component={RestaurantPage} />` 와 같이 설정하여 id값을 match 객체로 읽을 수 있도록 했는데,

문제는 레스토랑을 클릭해서 들어가면 제대로 된 레스토랑 정보가 나오지만, 직접 url을 쳐서 들어가면 에러가 나타납니다.

이 문제를 어떻게 해결해야 할까요 ?

Kwakcena committed:
index.html 수정

restaurants/1 과 같은 url 검색 시 id가 1번인 레스토랑 정보가 나오지 않은
문제가 있었다.

이를 해결하기 위해 다른 분의 코드리뷰를 참고하니 baseurl 관련 문제라는
말을 보았고, main.js 에서 /main.js로 수정하니 제대로 동작하였다.

그런데 이게 왜 이런건지 검색해봐도 잘 모르겠다...

Kwakcena committed:
refactor reducer test using fixtures

Kwakcena committed:
refactor reducer initialRestaurant

Kwakcena committed:
refactor RestaurantsContainer

Kwakcena committed:
refactor RestaurantContainer

Kwakcena commented:
다른 분의 코드리뷰를 통해서 `base path` 문제를 해결했습니다.

그런데 이유가 무엇때문인지는 아직 알지 못했습니다... 

왜 index.html에 `/main.js` 라고 하면 `restaurants/1` 을 직접 쳐서 들어갈 수 있고, `main.js` 라 하면 안되는건가요...?

참고할만한 자료가 있다면 부탁드립니다.

hannut91 commented:
그 부부는 절대경로 상대경로의 차이인 것 같아요

hannut91 commented on src/App.test.jsx:
```diff
@@ -1,34 +1,80 @@
+    ));
+  }
+
+  beforeEach(() => {
```
`dispatch.mockClear`를 해주셔야 의도치 않게 테스트가 동작하는 것을 방지할 수 있어요

hannut91 reviewed (changes requested):
`main.js`에 대한 문제는 `index.html`의 `head`에 
`<base href="/">`를 추가해서 해결할 수도 있습니다.

See also
  - https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base

Kwakcena committed:
dispatch.mockClear() 추가

Kwakcena committed:
<base> 태그 추가

<head> 태그 안에 <base> 태그를 사용하여 기준 절대경로 URL을 설정해주면
해당 웹페이지와 연결되는 링크 주소는 상대경로로 간소화하여 설정할 수
있다.

Kwakcena committed:
reducer setLoading 테스트

Kwakcena committed:
reducer loading test 수정

Kwakcena committed:
reducer loading test 구현

Kwakcena committed:
reducer refactor initialstate

초기 상태로 loading 을 추가

Kwakcena committed:
RestaurantContainer loading test

Kwakcena committed:
loading 페이지 구현

Kwakcena committed:
Refactor: 삼항연산자 사용

Kwakcena committed:
RestaurantContainer refactor

Kwakcena committed:
다시 제출

Kwakcena commented:
다시 제출하는 방법 말고 이미 제출한 PR에 대해 테스트를 다시 돌리는 방법을 모르겠어서 수정하고 제출합니다... 무슨 키워드로 찾아봐야 할지도 잘 모르겠어서...

이번에 추가한 것은 restaurant 정보를 출력할 때 loading page를 구현해봤습니다. 감사합니다.

hannut91 commented on src/RestaurantContainer.test.jsx:
```diff
@@ -0,0 +1,44 @@
+  }
+
+  context('when loading', () => {
+    it('is false show restaurant page', () => {
```
`renders restaurant`은 어떨까요?

hannut91 commented on src/RestaurantContainer.test.jsx:
```diff
@@ -0,0 +1,44 @@
+
+      expect(container).toHaveTextContent(restaurant.name);
+    });
+    it('is true show loading page', () => {
```
`renders loading`은 어떨까요?

hannut91 commented on src/RestaurantContainer.test.jsx:
```diff
@@ -0,0 +1,44 @@
+
+  context('when loading', () => {
+    it('is false show restaurant page', () => {
+      useSelector.mockImplementation((selector) => selector({
+        restaurant,
+        loading: false,
+      }));
```
각 컨텍스트마다 달라지는 것은 beforeEach로 표현하면 더 좋을 것 같아요.

```js
describe('RestaurantContainer', () => {
  function renderRestaurantContainer() {
    return render((
      <MemoryRouter>
        <RestaurantContainer />
      </MemoryRouter>
    ));
  }

  context('when loading is done', () => {
    beforeEach(() => {
      useSelector.mockImplementation((selector) => selector({
        restaurant,
        loading: false,
      }));
    });

    it('is false show restaurant page', () => {
      const { container } = renderRestaurantContainer();

      expect(container).toHaveTextContent(restaurant.name);
    });
  });
  
  context('when is loading', () => {
    beforeEach(() => {
      useSelector.mockImplementation((selector) => selector({
        restaurant,
        loading: true,
      }));
    })

    it('renders loading', () => {
      const { container } = renderRestaurantContainer();

      expect(container).toHaveTextContent('Loading...');
    });
  });
});
```

Kwakcena replied:
아하🤔 이부분을 어떻게 리팩토링 해야할까 고민했는데 저렇게 처리하면 좋을것 같군요 !

hannut91 commented on src/RestaurantPage.jsx:
```diff
@@ -0,0 +1,27 @@
+export default function RestaurantPage({ match: { params: { id } } }) {
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    dispatch(setLoading(true));
+    dispatch(setRestaurantId(id));
+    dispatch(loadRestaurant());
+  }, []);
```
`setLoading()`을 `loadRestaurant`액션에서 dispatch하는건 어떨까요?

`setRestaurantId`할 필요 없이 `loadRestaurant(id)`이렇게 하는건 어떨까요?

Kwakcena replied:
음... 다른 `load` 함수처럼 인자를 전달하지 않으려 맞췄는데 굳이 그럴 필요가 없었으려나요 ?

Kwakcena committed:
Refactor: RestaurantContainer

- 각각의 context마다 달라지는 것은 beforeEach로 표현한다.
- 테스트 명 수정

Kwakcena committed:
Refactor: Update loadRestaurant using setLoading

loadRestaurant에서 loading을 true로 만들고 데이터를 가져오는것이 완료 된
뒤 false를 만들도록 코드를 수정

Kwakcena committed:
Refactor: 테스트 코드와 구현 수정

Kwakcena committed:
actions test loadRestaurant

- setRestaurantId 를 할 필요 없이 id값을 loadRestaurant에 전달하여 바로
  레스토랑 정보를 얻도록 테스트

Kwakcena committed:
actions test 구현

Kwakcena committed:
Refactor: setRestaurantId 관련 코드 삭제

hannut91 commented on src/RestaurantContainer.test.jsx:
```diff
@@ -0,0 +1,49 @@
+    });
+  });
+
+  context('when is loading', () => {
```
`when loading`이 맞는 것 같구요 ㅎㅎ..제가 잘못썻네요

hannut91 commented on src/reducer.test.js:
```diff
@@ -31,9 +47,7 @@ describe('reducer', () => {
-      const regions = [
-        { id: 1, name: '서울' },
-      ];
+      const regions = [REGIONS[0]];
```
이거는 크게 의미가 없을 것 같아요. fixture를 사용하실꺼면 그냥 사용하면 좋을 것 같아요

```js
const state = reducer(initialState, setRegions(REGIONS));

expect(state.categories).toEqual(REGIONS);
```

hannut91 reviewed (approved):
변경사항이 있을 때 조금씩 변경해가면서 피해를 최소화하는 부분이 좋네요! 고생하셨습니다. 아샬님의 과제풀이를 보며 복습해주면 좋을 것 같아요!

hannut91 commented:
이번 과제를 통해 무엇을 배우셨나요? 더 배우고 싶은게 있으신가요?

Kwakcena committed:
Refactor: test 이름 수정

Kwakcena committed:
Refactor: 테스트 수정

Kwakcena commented:
# 내가 한 것(Fact)

- 레스토랑 상세 페이지 구현하기.
- `react-router-dom` 을 이용해서 리액트를 이용해 라우팅하기.
- 데이터를 `fetch` 하는 동안 화면에 보여 줄 `loading` 페이지 구현.
- `<base>` 태그를 사용하여 baseURL 문제를 해결.

# 느낀 점(Feelings)

- 과제를 수행하면서 라우팅 뿐만 아니라 저번시간에 배운 비동기를 활용한 과제라는 점에서 복습도 할 수 있었고 , 라우팅도 적용할 수 있었다.
- 수행한 과제 중에서 가장 재미있었다.
- url에 직접 `restaurants/1` 과 같이 검색하여 레스토랑 정보를 가져올 때 절대경로와 상대경로의 차이로 인해 발생하던 에러를 해결함에 있어 `<base>` 태그를 이용해 해결했는데, 다른 분은 웹팩의 설정을 통해 해결했다는 리뷰를 보았다. 웹팩 설정을 잘 할줄 알아야 겠다는 느낌을 받았다.
- 이번에 수행한 TDD는 사이클을 나름 잘 지킨 것 같아 뿌듯하다.
- 영어로 테스트를 작성하고 싶은데 참 어렵다. 😭
- 과제의 첫 코드로 주어진 테스트 코드를 보고 감탄했다.

# 배운 점(Finding)

1. `<base>` 태그를 사용하면 기준 절대경로를 정할 수 있다. 이후로는 웹 페이지와 연결되는 링크 주소를 상대경로로 간소화 할 수 있다.
2. `<a>` 태그를 사용했을 때와 `<Link>` 를 사용할 때의 차이점을 알 수 있었다. 페이지 자체의 리로딩 없이 하려면 `<Link>` 를 써야 한다.
3. `<Route>` 컴포넌트의 `component` prop으로 넣어준 컴포넌트에 `match, location, history` 라는 3개의 prop을 넘겨준다는 것을 배웠다.
4. `<Route path="restaurants/:id" ...>` 에서 경로 부분의 `:id` 는 URL 파라미터를 정의할 때 사용하는 React Router의 문법임을 알 수 있었다. URL 파라미터가 포함된 경우 패턴 매칭에 의해 해당 파라미터는 `component` prop로 전달한 컴포넌트에서 `match.params.id` 와 같이 읽을 수 있다.
5. `<Route>` 에서 정확히 일치하는 URL에 대해서만 라우팅 하고 싶다면 `exact` 를 써야 함을 배웠다.

# 내가 할 것(Affirmation)

- 아샬님 풀이 강의 보면서 복습하기
- 지금은 간략하게 배운점을 썻지만 이걸 자세히 써서 블로그에 써야 하는데... 안쓰는 나.
- `api` 를 테스트 하던 코드에 대해 자세히 알아보자.

hannut91 commented:
블로그에 정리한걸 기대해봅니다!
